app/src/ts.app.js

```diff
@@ -0,0 +1,168 @@
+	 */
+	constructor(appIds) {
+		this.appIds = appIds;
+		this.events = new Map();
+	}
+
+	/**
+	 * Listen to message and call handler.
+	 * @param {string} key The key/subject of the event.
+	 * @param {function} callback The handler of the event.
+	 * @return {Listener} Listener object
+	 */
+	on(key, callback) {
+		if (!key || !callback) {
+			console.warn(`We need two parameters key and callback`);
+			return this;
+		}
+		if (this.events.has(key)) {
+			console.warn(`You have listened the message of ${key}`);
+			return this;
+		}
+		const handler = e => {
+			if (typeof e.data !== 'string' || !e.data.includes(BROADCAST_PREFIX) || !e.data.includes(key)) {
+				return this;
+			}
+			let content = e.data.replace(BROADCAST_PREFIX, '');
+			let data = JSON.parse(content);
+			if (data.key !== key) {
+				return this;
+			}
+			if (callback) {
+				callback(data);
+			}
+		}
+		window.addEventListener('message', handler);
+		this.events.set(key, handler);
+		return this;
+	}
+
+	/**
+	 * Stop listen to message.
+	 * @param {string} key The key/subject of the event.
+	 * @return {Listener} Listener object
+	 */
+	off(key) {
+		if (!key) {
+			console.warn(`We need the parameter of key`);
+			return this;
+		}
+		if (!this.events.has(key)) {
+			console.warn(`Can\'t find the message of ${key}`);
+			return this;
+		}
+		const handler = this.events.get(key);
+		window.removeEventListener('message', handler);
+		this.events.delete(key);
+		return this;
+	}
+
+	/**
+	 * Temporarily stop receiving events, but keep all handlers.
+	 * @param {string} key The key/subject of the event.
+	 * @return {Listener} Listener object
+	 */
+	pause(key) {
+		if (!key) {
+			console.warn(`We need the parameter of key`);
+			return this;
+		}
+		if (!this.events.has(key)) {
+			console.warn(`Can\'t find the message of ${key}`);
+			return this;
+		}
+		const handler = this.events.get(key);
+		window.removeEventListener('message', handler);
+		return this;
+	}
+
+	/**
+	 * Resume a listener after a pause().
+	 * @param {string} key The key/subject of the event.
+	 * @return {Listener} Listener object
+	 */
+	resume(key) {
+		if (!key) {
+			console.warn(`We need the parameter of key`);
+			return this;
+		}
+		if (!this.events.has(key)) {
+			console.warn(`Can\'t find the message of ${key}`);
+			return this;
+		}
+		const handler = this.events.get(key);
+		window.addEventListener('message', handler);
+		return this;
+	}
+
+	/**
+	 * Stop listening to all registered handlers.
+	 */
+	clear() {
+		for (let key of this.events.keys()) {
+			this.off(key);
+		}
+		return this;
+	}
+}
+
+// Private .................................................................
+
+/**
+ * Encode broadcast to be posted.
+ * @param {array|string} appIds List of apps to receive the message, supports glob
+ * @param {string} key The key/subject of the event
+ * @param {object} data Data to be sent with the event, has to support the structured clone algorithm
```

Comment:
I don't think the structured clone algorithm matters much since we stringify it anyways, in fact there's a chance that it must be a string because otherwise failure, although I didn't look it up. It has to be a string anyways in order to support Internet Exploder, as far as I know.